exploration/sigils.md

```diff
@@ -0,0 +1,417 @@
+
+### P02: It should be possible to avoid repetition of open/close palceholders in the message body by assigning them to local variables.
+
+We can already do this for standalone placeholders.
+
+	// now
+	let $x = {+html opt=val}
+	{{$x}Hello{-html}}
+
+	// better?
+	let $x = {:html opt=val}
+	{{+ $x}Hello{-html}}
+
+	// better? But: what about opt=val being passed to "close"?
+	let $x = {:html opt=val}
+	{{+ $x}Hello{- $x}}
+
+## Open Questions
+
+### Q01: Should it be possible to specify open and close properties on *placeholders* without annotations?
+
+This follows from P02 above:
+
+	let $x = {:html opt=val}
+	{{+ $x}Hello{-html}}
+	
+Is it also useful to allow open/close literals?
+If formatted parts carry the open/close information, they could still be used by higher abstraction layers.
+However, it would impair tooling.
+
+	// without annotations, tooling doesn't know anything about "a"
+	{{+ a}text{- a}}
+
+> [!IMPORTANT]
+> If so:
+> Open/close is not a property of the annotation.
+> It is at most a property of the expression, which can comprise just the operand.
+> Or perhaps even a property of the placeholder (see below).
+
+### Q02: Should it be possible to specify open/close on expressions used in local variable declarations?
+
+R08 is about making open/close invalid in selectors.
+Perhaps also forbid open/close in local variable declarations and only allow them in placeholders inside patterns?
+
+	// Local variables must be standalone annotations.
+	let $x = {:html opt=val}
+
+	// Syntax TBD; assuming open/close is a property of the placeholder.
+	{{+ $x}Hello{- $x}}
+
+	// Or perhaps repeat the annotation?
+	{{$x +html}Hello{$x -html}}
+
+> [!IMPORTANT]
+> If so:
+> Open/close is not a property of the annotation.
+> It is at most a property of the expression, which can comprise just the operand.
+> Or perhaps even a property of the placeholder (see below).
+
+### Q03: Should it be possible to change the open/close role once assigned?
+
+Assuming we allow Q02 above.
+We can require a data model error be emitted when this happens.
+
+	let $x = {em +html}
+	{{$x -html}}
+
+### Q04: Should formatting and selectors annotations have different syntax?
+
+	See [#260](https://github.com/unicode-org/message-format-wg/issues/260).
+
+### Q05: Do we need namespaced local variables?
+
+	See [#403](https://github.com/unicode-org/message-format-wg/issues/403).
+
+## Alternatives for the syntax
+
+Once we answer the open questions, we can design the sigils and the syntax for open/close.
+Some options.
+
+### Function prefix
+
+Single char sigils in front of function names.
+
+	{img :html}
+	{img +html}
+	{img -html}
+
+* at odds with P01; is the function called `html` or `:html`?
+
+Double char sigils to group similar sigils under a common introducer.
+
+	{img :html}
```

Comment:
I think that technically these should be
:html
::html
/:html
+:html
and so on.

`:html` means function.
And if open/close/etc are not function attributes, they they belong "outside" the `:html`

Not saying it is agreement yet that these are function attributes.
But if we agree or not, it affects the order of the sigils / sub-sigils.

---

In general I think that if open / close are function attributes they can be represented in the options bag: `{a :html type:closed}`, there is no need to single them out, except for convenience of typing.

But if the same flag shows again and again across many functions, and can be used (by humans and linters) without any understanding or access to the registry, then I argue it is not a function attribute.
I can tell that `"...{foo -bar}...{foo +bar}..."` is wrong even if I have no clue what `bar` is / does.